# Using synthetic chromosome controls to evaluate the sequencing of difficult regions within the human genome

## ABSTRACT

  **Background**: Next-generation sequencing (NGS) can identify mutations in the human genome that cause disease and has been widely adopted in clinical diagnosis. However, the human genome contains many polymorphic, low complexity, and repetitive regions that are difficult to sequence and analyse. Despite their difficulty, these regions include many clinically-important sequences that can inform the treatment of human diseases and improve the diagnostic yield of NGS. 
  **Results**: To evaluate the accuracy by which these difficult regions are analysed with NGS, we built an in silico decoy chromosome, along with corresponding synthetic DNA reference controls, that encode difficult and clinically-important human genome regions, including repeats, microsatellites, HLA genes and immune-receptors. These controls provide a known ground-truth reference against which to measure the performance of diverse sequencing technologies, reagents, and bioinformatic tools. Using this approach, we provide a comprehensive evaluation of short- and long-read sequencing instruments, library preparation methods, and software tools, and identify the errors and systematic bias that confound our resolution of these remaining difficult regions. 
  **Conclusions**: This study provides an analytical validation of diagnosis using NGS in difficult regions of the human genome and highlights the challenges that remain to resolve these difficult regions.

## CODE AVAILABILITY

Anaquin, the toolkit used in the sequins analyses, is freely available at https://www.sequinstandards.com and all scripts used to perform statistical analyses and generate plots can be found at https://github.com/almreis/Benchmark_ChrQ.
